### ISSUE TITLE
Call to signal.signal( ) in application.py fails due to signal.getsignal( ) returning None.

### ISSUE THREAD

**Issue by john-odonovan** [prompt-toolkit/python-prompt-toolkit#1108]
Hi,

I've been integrating ipython into an application for a number of years, but recently found a failure which is coming from prompt-toolkit. The issue on only in v3 of the code. The problem occurs at 
 prompt-toolkit/python-prompt-toolkit/application.py:727 (in the master)

    signal.signal(signal.SIGWINCH, previous_winch_handler)

previous_winch_handler was assigned at line 691. 

    previous_winch_handler = signal.getsignal(signal.SIGWINCH)

This call returned None, and then the call on line 727 fails, since None is not an acceptable value for `signal.signal()`.

Just prior to invoking the Python interpreter in my C++ application, I use `signal( SIGWINCH, SIG_DFL )` to remove any installed signal handler from the C++ code. I also tried SIG_IGN, but to no avail, `getsignal()` still returns None. One workaround was to execute the following in my Python code before initializing IPython, and subsequently calling prompt-toolkit.

    signal.signal( signal.SIGWINCH, signal.SIG_DFL ) 

So my question is; should the prompt-toolkit code account for the fact that `getsignal()` could return None? Or is there something better that I should do in my application? Or is my current workaround the way forward?

Regards,
  John

**Comment by john-odonovan** [comment]
Hi,

Previously I had the call to `signal( SIGWINCH, SIG_DFL );` before the call to `PyRun_SimpleString()`, but after the call to `Py_Initialize();`. If I move it to before the call to `Py_Initialize();`, then prompt-toolkit will also work, so my two workarounds are 

1. In C++ code, `signal( SIGWINCH, SIG_DFL )` before the call to `Py_Initialize()`.
2. In Python code, `signal.signal( signal.SIGWINCH, signal.SIG_DFL )` before starting IPython/prompt-toolkit.

Question still remains; should this be fixed in prompt-toolkit, or should an application embedding calls to prompt-toolkit use one of the approaches above?

Regards,
  John

**Comment by jonathanslenders** [comment]
@john-odonovan,

Would this solve it? https://github.com/prompt-toolkit/python-prompt-toolkit/pull/1110/files

I've seen a situation before where we receive None too, but I don't remember exactly when that happened. Could it be that a signal handler was set by C code instead of Python code?

I guess we can restore to the default in this situation.

**Comment by john-odonovan** [comment]
@jonathanslenders 

Yes, this will resolve the issue. The more complex alternative would be to have special handling of the None return from getsignal and in that scenario call the OS sigaction function to get the existing handler, and then restore this later. But, in general there are so many modules that could manipulate signal handlers, that it would not be worth it to have a more complex implementation in prompt-toolkit. I think that your proposal will be sufficient.

**Comment by jonathanslenders** [comment]
Ok, thanks for the feedback! I've merged it like that.

**Comment by john-odonovan** [comment]
Thanks Jonathan.
